src/robot/output/jsonlogger.py

```diff
@@ -0,0 +1,241 @@
+
+    def set_log_level(self, level):
+        return self._log_message_is_logged.set_level(level)
+
+    def message(self, msg):
+        if self._error_message_is_logged(msg.level):
+            self._errors.append(msg)
+
+    def log_message(self, msg):
+        if self._log_message_is_logged(msg.level):
+            self._write_message(msg)
+
+    def _write_message(self, msg):
+        message = self._create_message(msg)
+        if self._current_item == 'test':
+            self._current_test['msg'] = message
+        elif self._current_item == "keyword":
+            if 'msgs' not in self._current_keyword:
+                self._current_keyword['msgs'] = list()
+            self._current_keyword['msgs'].append(message)
+
+    def _create_message(self, msg):
+        message = {
+            'msg': msg.message,
+            'level': msg.level,
+            'timestamp': msg.timestamp or None,
+            'html': msg.html is not None or None
+        }
+        return self._format_data(message)
+
+    def start_keyword(self, kw):
+        # If there is an "open" keyword, this will be placed inside its keywords
+        if self._current_keyword:
+            # Push this onto the stack
+            self._item_stack.append(self._current_keyword)
+        self._current_keyword = {
+            'name': kw.kwname,
+            'lib': kw.libname,
+            'type': kw.type if kw.type != 'kw' else None,
+            'doc': kw.doc,
+            'tags': [unic(t) for t in kw.tags],
+            'args': [unic(a) for a in kw.args],
+            'assign': [var for var in kw.assign],
+            'destination': copy.deepcopy(self._current_item)
+        }
+
+        # Mark the type of item
+        self._current_item = 'keyword'
+
+    def end_keyword(self, kw):
+        # Check if we have an item in progress
+        if not self._current_keyword:
+            # If not, then the keywords have been processed for the item on the stack
+            self._current_keyword = self._item_stack.pop()
+
+        # Add the rest of the information
+        if kw.timeout:
+            self._current_keyword['timeout'] = unic(kw.timeout)
+        self._current_keyword['status'] = self._create_status(kw)
+
+        # Remove the "destination" from the dictionary
+        destination = self._current_keyword.pop('destination')
+        # Format the keyword
+        self._current_keyword = self._format_data(self._current_keyword)
+        # To process the keyword we must have pushed something to the item stack (or have an active suite/test)
+        if destination == 'suite':
+            parent_item = self._current_suite
+        elif destination == 'test':
+            parent_item = self._current_test
+        else:
+            parent_item = self._item_stack[-1]
+
+        if 'kw' not in parent_item:
+            parent_item['kw'] = list()
+        parent_item['kw'].append(self._current_keyword)
+        # Mark the current keyword as being none
```

Comment:
not sure these 2 comments have much added value. The code is clear enough.

Reply:
I've removed the comments